### PR TITLE
Re-structuring `dcos-cni` repo to be more consistent with golang idioms

### PR DESCRIPTION
**Pull request by asridharan**
These changes address @jdef's comments on organizing the code more inline with Golang idioms.

---
This change is [Reviewable](https://reviewable.io/reviews/dcos/dcos-cni/4)

### CODE REVIEW

**Comment by asridharan** [comment]
@jdef could you take a look at this PR. Trying to address most of your earlier comments on the repo.

File: pkg/l4lb/config.go [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+	"fmt"
+
+	"github.com/containernetworking/cni/pkg/types"
+
+	"github.com/dcos/dcos-cni/pkg/minuteman"
+)
+
+type NetConf struct {
+	types.NetConf
+	Spartan   bool                   `json:"spartan, omitempty"`
```
Comment (jdef) [inline pkg/l4lb/config.go]: could this be more clearly named `SpartanEnabled`? (if that's what it really means...)

File: pkg/l4lb/config.go [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+	conf.Delegate["cniVersion"] = conf.CNIVersion
+	conf.Delegate["args"] = conf.Args
+
+	delegateConf, err = json.Marshal(conf.Delegate)
+	if err != nil {
+		err = fmt.Errorf("failed to marshall the delegate configuration: %s", err)
+		return
+	}
+
+	_, ok := conf.Delegate["type"]
```
Comment (jdef) [inline pkg/l4lb/config.go]: DRY: save this value and use it later?

File: pkg/l4lb/config.go [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+	if !ok {
+		err = fmt.Errorf("type field missing in delegate network: %s", conf.Delegate["name"])
+		return
+	}
+
+	delegatePlugin, ok = conf.Delegate["type"].(string)
```
Comment (jdef) [inline pkg/l4lb/config.go]: DRY: you already looked up the value of `conf.Delegate["type"]`

File: pkg/spartan/plugin.go [hunk 4]
```diff
@@ -80,53 +87,53 @@ func CniAdd(args *skel.CmdArgs) error {
...
 	ipamResult, err := ipam.ExecAdd(Config.IPAM.Type, spartanNetConf)
 	if err != nil {
-		return err
+		return Error(err.Error())
 	}
 
 	result, err := current.NewResultFromResult(ipamResult)
 	if err != nil {
-		return err
+		return Error(err.Error())
```
Comment (jdef) [inline pkg/spartan/plugin.go]: dubious. just return `err`

File: pkg/spartan/plugin.go [hunk 5]
```diff
@@ -80,53 +87,53 @@ func CniAdd(args *skel.CmdArgs) error {
...
 	if err != nil {
-		return fmt.Errorf("failed to marshall the `spartan-network` IPAM configuration: %s", err)
+		return Error(fmt.Sprintf("failed to marshall the `spartan-network` IPAM configuration: %s", err))
 	}
 
 	// Run the IPAM plugin for the spartan network.
 	ipamResult, err := ipam.ExecAdd(Config.IPAM.Type, spartanNetConf)
 	if err != nil {
-		return err
+		return Error(err.Error())
```
Comment (jdef) [inline pkg/spartan/plugin.go]: dubious. just return `err`

File: pkg/spartan/plugin.go [hunk 6]
```diff
@@ -80,53 +87,53 @@ func CniAdd(args *skel.CmdArgs) error {
...
+		return Error("Expecting a single IPv4 address from IPAM")
 	case result.IPs[0].Address.IP.To4() == nil:
-		return errors.New("Expecting a IPv4 address from IPAM")
+		return Error("Expecting a IPv4 address from IPAM")
 	}
 
 	hostVethName, err := setupContainerVeth(args.Netns, Config.Interface, 0, *result, IPs)
 	if err != nil {
-		return err
+		return Error(err.Error())
```
Comment (jdef) [inline pkg/spartan/plugin.go]: dubious. just return `err`

File: pkg/spartan/plugin.go [hunk 7]
```diff
@@ -135,11 +142,11 @@ func CniAdd(args *skel.CmdArgs) error {
 func CniDel(args *skel.CmdArgs) error {
 	spartanNetConf, err := json.Marshal(Config)
 	if err != nil {
-		return fmt.Errorf("failed to marshall the `spartan-network` IPAM configuration: %s", err)
+		return Error(fmt.Sprintf("failed to marshall the `spartan-network` IPAM configuration: %s", err))
 	}
 
 	if err = ipam.ExecDel(Config.IPAM.Type, spartanNetConf); err != nil {
-		return err
+		return Error(err.Error())
```
Comment (jdef) [inline pkg/spartan/plugin.go]: dubious. just return `err`

**Comment by jdef** [comment]
left a few suggestions

On Thu, Apr 27, 2017 at 8:04 PM, Avinash Sridharan <notifications@github.com

**Comment by asridharan** [comment]
Review status: 0 of 9 files reviewed at latest revision, 7 unresolved discussions.

---

*[pkg/l4lb/config.go, line 14 at r1](https://reviewable.io:443/reviews/dcos/dcos-cni/4#-Kilrhckn57MNDxUuY8k-r1-14:-KjAp3UWLnZYqPNgcJC1:b-ywhtu0) ([raw file](https://github.com/dcos/dcos-cni/blob/e08926280e4e62324a2d3a8af03bc8897d40cb7d/pkg/l4lb/config.go#L14)):*
<details><summary><i>Previously, jdef (James DeFelice) wrote…</i></summary><blockquote>

could this be more clearly named `SpartanEnabled`? (if that's what it really means...)
</blockquote></details>

Seems a bit verbose? The fact that Spartan is true|false implies its enabled/disabled. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dcos/dcos-cni/4)*

**Comment by jdef** [comment]
Review status: 0 of 9 files reviewed at latest revision, 8 unresolved discussions.

---

*[pkg/l4lb/config.go, line 14 at r1](https://reviewable.io:443/reviews/dcos/dcos-cni/4#-Kilrhckn57MNDxUuY8k-r1-14:-KjBNAQT4dxmvRehkefp:b-tiuy0v) ([raw file](https://github.com/dcos/dcos-cni/blob/e08926280e4e62324a2d3a8af03bc8897d40cb7d/pkg/l4lb/config.go#L14)):*
<details><summary><i>Previously, asridharan (Avinash Sridharan) wrote…</i></summary><blockquote>

Seems a bit verbose? The fact that Spartan is true|false implies its enabled/disabled. 
</blockquote></details>

I think I left another comment on the commit in GH. Bottom line: I won't reject the changeset for this, but without a comment `Spartan` is not self-explanatory. A better name would be great.

---

*[pkg/l4lb/config.go, line 23 at r2](https://reviewable.io:443/reviews/dcos/dcos-cni/4#-KjBNYmoYRuRShzAv5uX:-KjBNYmoYRuRShzAv5uY:b-q8dtfw) ([raw file](https://github.com/dcos/dcos-cni/blob/f5949365b1dfc10d8f8be229ea81a3537388dc13/pkg/l4lb/config.go#L23)):*
> ```Go
> func NewNetConf() *NetConf {
> 	conf := &NetConf{
> 		Spartan: true,
> ```

FWIW it's always useful when the zero-value of a struct has meaning (like "the sensible default"). In this case, the zero-val means the opposite of the thing produced by this struct because the bools are false in a zero-val. It may be more convenient to invert the flags a bit in the structs: `SpartanDisabled` and `Minuteman.Disabled`. That way the zero val `&NetConf{}`, to be returned by this func, makes sense as a good default.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dcos/dcos-cni/4)*

File: cmd/l4lb/l4lb.go [hunk 8]
```diff
@@ -60,12 +60,8 @@ func cmdAdd(args *skel.CmdArgs) error {
 		return fmt.Errorf("failed to invoke delegate plugin %s: %s", delegatePlugin, err)
 	}
 
-	if !conf.Spartan && conf.Minuteman == nil {
-		return fmt.Errorf("at least one of minuteman or spartan CNI options need to be enabled for this plutin")
-	}
-
 	fmt.Fprintln(os.Stderr, "Spartan enabled:", conf.Spartan)
```
Comment (jdef) [inline cmd/l4lb/l4lb.go]: move this log line into the following `if`?

Comment (jdef) [reply cmd/l4lb/l4lb.go]: also, why use `fmt.Fprintln` vs the `log` package?

File: cmd/l4lb/l4lb.go [hunk 9]
```diff
@@ -60,12 +60,8 @@ func cmdAdd(args *skel.CmdArgs) error {
 		return fmt.Errorf("failed to invoke delegate plugin %s: %s", delegatePlugin, err)
 	}
 
-	if !conf.Spartan && conf.Minuteman == nil {
```
Comment (jdef) [inline cmd/l4lb/l4lb.go:110]: why is there no need for this check anymore?

**Review by jdef — APPROVED** [review]

File: cmd/l4lb/l4lb.go (hunk 9 above)
Comment (jdef) [reply cmd/l4lb/l4lb.go:110]: still wondering about this...

**Pull request merged**